[knowledge-base] Add Like & Favorite to Guide Detail Page

yizou275 opened the pull request:
This PR is for implementing user's "like" or "favorite" reaction to a guide, as described by the user stories.


## Description
The implemented feature includes the following:
1. Display "like" and "favorite" buttons for logged in users and show their current state of whether they like or favorite this guide (hide the buttons for unlogged in users)
2. Allow user interaction, when they click on "like", this toggles their state (if initially liked, now unliked; if initially unliked, now liked); same applies to "favorite" button
3. The toggling of state will manifest on the appearance of the button and on the like counts and favorite counts on top of the page.

## File Changes
Specifically, the files I change are:
- `views.py`: added two views, `LikeUnlikeGuide` and `FavUnfavGuide`, that return JSON responses on whether the final state is liked/favorited, and the number of counts after the change
- `urls.py`: add urls for the corresponding views
- `guide_detail.html`: add buttons and update on like/favorite counts
- `like-favorite.js`: allow user interaction with the button that changes the page display without having to reload the page
- `single-guide.css`: add heart and star styles for the buttons

## Testing
To test it, first pull this branch, load the fixture according to `knowledge-base/fixtures/README.md`, and then runserver.
1. Log in as Admin. Go to "Guide for Go Game" page (http://127.0.0.1:8000/knowledge-base/guides/2). Confirm it looks like the following. Note the heart and star buttons below, and confirm they match the current state of the Guide object.
<img width="1377" alt="Screenshot 2025-05-20 at 9 54 57 PM" src="https://github.com/user-attachments/assets/d0b2d604-df49-493a-b3f4-d42f45b4a3d2" />
2. Try to like it and unfavorite it. Confirm the button changes accordingly, counts on top of the page changes accordingly, and go to admin interface to confirm the guide object's manytomanyfield is updated correctly. Updated page look like the following.
<img width="1423" alt="Screenshot 2025-05-20 at 9 57 29 PM" src="https://github.com/user-attachments/assets/98f36c1c-447f-4596-babc-42fba194966b" />

3. Reload the page and confirm this is still the case.
4. Log out and go to the same page. Confirm you can no longer see the two buttons now.

> AI reference: Get some of the CSS codes by modifying the example code from ChatGPT.

Commit:
[knowledge-base] WIP: guide like

- add view and url for like, add basic template
- will further work on frontend appearance of the like feature
- then do the same thing for favorites

For: #910

Commit:
[knowledge-base] WIP: like guide feature

- change implementation to JSON and Javascript so the page doesn't need to
reload when someone likes a guide
- will do the same thing for favorite

For: #910

Commit:
[knowledge-base] add favorite

- add url, view, html/css templates and js for favorites as well
- fix minor issues on likes

For: #910

Commit:
[knowledge-base] hide like/favorite for unlogged in users

For: #910

yizou275 committed:
Merge branch 'dev' into knowledge-base/viewer/like-and-favorite-guides

yizou275 commented:
Also, note that now the buttons appear on the bottom of the page, because the table of content hasn't been rendered yet.

mrollinswaterman commented:
Having a slight issue, everything basically works fine, but my Like and Favorite buttons aren't colored like yours are, no matter what I do. Another small edit, not sure how hard this would be, but having the "Favorite" switch to "Unfavorite" if the user already has the guide favorited would be a nice touch. The coloring of icons bug comes first though.

mrollinswaterman committed:
Merge branch 'dev' into knowledge-base/viewer/like-and-favorite-guides

mrollinswaterman reviewed (approved):
In regards to my previous issue, it was actually a problem with my browser not updating the CSS file properly and using the cache instead. After I bypassed the cache, all functionality was online! Great work Yi!

mrollinswaterman committed:
Merge branch 'dev' into knowledge-base/viewer/like-and-favorite-guides

Commit:
[knowledge-base] Modify like/fav button text

- modify the text so that when it's already liked/favorited, the text
displays "unlike"/"unfavorite"

For: #910

Commit:
Merge branch 'knowledge-base/viewer/like-and-favorite-guides' of https://github.com/uchicago-cs/chigame into knowledge-base/viewer/like-and-favorite-guides

yizou275 committed:
Merge branch 'dev' into knowledge-base/viewer/like-and-favorite-guides

yizou275 commented:
Thanks Miles! Regarding your second point, I changed the text of the button so that it displays "unlike" when a user originally "likes" it. Same applies to "unfavorite".

jacqsmei committed:
Merge branch 'dev' into knowledge-base/viewer/like-and-favorite-guides

jacqsmei reviewed (approved):
nice work! cool feature to have

jacqsmei commented:
Issue Score: Excellent

Comments:
Really well written PR, nice work! And glad you both were able to figure out the css thing

<!-- {"score": "excellent", "minor": false} -->
